### ISSUE TITLE
Proper accuracy calculation on multi-gpu with DistributedDataParallel with unevenly divisible data size

### ISSUE THREAD

**Issue by myron** [Project-MONAI/MONAI#5357]
How do we calculate proper DiceMetric accuracy (or any other metric) on multi-gpu DistributedDataParallel  when dataset is not evenly divisible by number of gpus?

For example on the last epoch, on 8 gpu machine, we have only 5 images, so 3 gpus will be idle .  So we need to records results only from 5gpus, and ignore other 3gpus.
Monai has a subclass of [DistributedSampler](https://github.com/Project-MONAI/MONAI/blob/c319ca8ff31aa980a045f1b913fb2eb22aadb080/monai/data/samplers.py) , which seems to account for non-even division using parameter even_divisible but if I set it to False, the DistributedDataParallel freezes since some GPUs now  receive no data.

Based on the tutorial code from [here](https://github.com/Project-MONAI/tutorials/blob/b429c883dccde3bfaa16c7c42fc2a35a8157fc3f/acceleration/distributed_training/unet_evaluation_ddp.py#L136) 

it creates a dataset from 13 images, and if running on 8 gpus it will freeze. 
It will freeze because of any syncs between gpus in this case the line `current_accuracy = dice_metric.aggregate().item()` is the problem (added to demonstrate the issue), but any `torch.distributed.barrier()` would also freeze it.  DDB expects the same code path for all processes and with DistributedSampler.even_divisible=False, it'll create different path in the last epoch (and it will only work if no sync is called, which is not convenient, as we need to calculate intermediate results or some other reduction calls). 

Here is the example code:
```
import argparse
import os
from glob import glob

import nibabel as nib
import numpy as np
import torch
import torch.distributed as dist
from torch.nn.parallel import DistributedDataParallel

import monai
from monai.data import DataLoader, Dataset, create_test_image_3d, DistributedSampler, decollate_batch
from monai.inferers import sliding_window_inference
from monai.metrics import DiceMetric
from monai.transforms import Activations, AsChannelFirstd, AsDiscrete, Compose, LoadImaged, ScaleIntensityd, EnsureTyped, EnsureType


def evaluate(args):
    if args.lo
```cal_rank == 0 and not os.path.exists(args.dir):
        # create 16 random image, mask paris for evaluation
        print(f"generating synthetic data to {args.dir} (this may take a while)")
        os.makedirs(args.dir)
        # set random seed to generate same random data for every node
        np.random.seed(seed=0)
        for i in range(16):
            im, seg = create_test_image_3d(128, 128, 128, num_seg_classes=1, channel_dim=-1)
            n = nib.Nifti1Image(im, np.eye(4))
            nib.save(n, os.path.join(args.dir, f"img{i:d}.nii.gz"))
            n = nib.Nifti1Image(seg, np.eye(4))
            nib.save(n, os.path.join(args.dir, f"seg{i:d}.nii.gz"))

    # initialize the distributed evaluation process, every GPU runs in a process
    dist.init_process_group(backend="nccl", init_method="env://")

    images = sorted(glob(os.path.join(args.dir, "img*.nii.gz")))
    segs = sorted(glob(os.path.join(args.dir, "seg*.nii.gz")))
    val_files = [{"img": img, "seg": seg} for img, seg in zip(images, segs)]

    # define transforms for image and segmentation
    val_transforms = Compose(
        [
            LoadImaged(keys=["img", "seg"]),
            AsChannelFirstd(keys=["img", "seg"], channel_dim=-1),
            ScaleIntensityd(keys="img"),
            EnsureTyped(keys=["img", "seg"]),
        ]
    )

    # create a evaluation data loader
    val_files = val_files[:13]

    val_ds = Dataset(data=val_files, transform=val_transforms)
    # create a evaluation data sampler
    val_sampler = DistributedSampler(dataset=val_ds, even_divisible=False, shuffle=False)
    # sliding window inference need to input 1 image in every iteration
    val_loader = DataLoader(val_ds, batch_size=1, shuffle=False, num_workers=2, pin_memory=True, sampler=val_sampler)
    dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)
    post_trans = Compose([EnsureType(), Activations(sigmoid=True), AsDiscrete(threshold=0.5)])
    # create UNet, DiceLoss and Adam optimizer
    # print(args)
    device = torch.device(f"cuda:{args.local_rank}")
    torch.cuda.set_device(device)
    model = monai.networks.nets.UNet(
        spatial_dims=3,
        in_channels=1,
        out_channels=1,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
    ).to(device)
    # wrap the model with DistributedDataParallel module
    model = DistributedDataParallel(model, device_ids=[device])
    # config mapping to expected GPU device
    # map_location = {"cuda:0": f"cuda:{args.local_rank}"}
    # load model parameters to GPU device
    # model.load_state_dict(torch.load("final_model.pth", map_location=map_location))


    model.eval()
    with torch.no_grad():
        i=0
        for val_data in val_loader:
            val_images, val_labels = val_data["img"].to(device), val_data["seg"].to(device)
            # define sliding window size and batch size for windows inference
            roi_size = (96, 96, 96)
            sw_batch_size = 4
            val_outputs = sliding_window_inference(val_images, roi_size, sw_batch_size, model)
            val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]

            d= dice_metric(y_pred=val_outputs, y=val_labels)
            current_accuracy = dice_metric.aggregate().item() # this line requires sync across gpus, and will freeze the system
            print(args.local_rank, 'iter', i, 'dice', d, current_accuracy, "\n")

            i = i + 1


        print(args.local_rank, 'aggregate', len(dice_metric.get_buffer()))

        metric = dice_metric.aggregate().item()
        dice_metric.reset()

        if dist.get_rank() == 0:
            print("evaluation metric:", metric)
        dist.destroy_process_group()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-d", "--dir", default="./testdata", type=str, help="directory to create random data")
    # must parse the command-line argument: ``--local_rank=LOCAL_PROCESS_RANK``, which will be provided by DDP
    parser.add_argument("--local_rank", type=int)
    args = parser.parse_args()

    evaluate(args=args)


# usage example(refer to https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py):

# python -m torch.distributed.launch --nproc_per_node=8 --nnodes=1 --node_rank=0 --master_addr="127.0.0.1" --master_port=12345 unet_evaluation_ddp.py -d ./tmp

if __name__ == "__main__":
    main()

**Comment by myron** [comment]
> Thanks for raising the ticket, I will try to find a machine and reproduce your issue. BTW, I think here "For example on the last epoch, on 8 gpu machine, we have only 5 images" you mean "last iteration".
> 
> Thanks.

yes, last iteration. thank you.

**Comment by Nic-Ma** [comment]
Hi @myron ,

I can reproduce your issue locally with the test program.
And after analysis, I found the issue is caused by:
```py
... ...
current_accuracy = dice_metric.aggregate().item() # this line requires sync across gpus, and will freeze the system
print(args.local_rank, 'iter', i, 'dice', d, current_accuracy, "\n")
... ...
```
You should not call `dice_metric.aggregate()` during the for loop.
When removed these 2 lines, every thing goes fine.
Could you please help verify it?

Thanks in advance.

**Comment by myron** [comment]
this is correct. I specifically added that line to demonstrate the issue. 

During the loop we want to print the intermediate result (average dice at that iteration).  Otherwise during long loop (in validation) we give no feedback (no intermediate results).   

In the current recommended workflow with it seems to be not possible.

**Comment by Nic-Ma** [comment]
Hi @myron ,

The reason is that when you do `aggregate`, it will do `all-gather` across ranks, if not all the ranks can run the same logic, it will hang to wait. I think that's CUDA related problem.

Thanks.

**Comment by myron** [comment]
Correct, but it's not the Cuda issue. Monai added a subclass of DistributedSampler with a new option even_divisible to generate datasets of non-equal length, hence the loops over such datasets will execute different number of steps in multi-Gpu run, and any reduction/sync will deadlock. our own metrics inside of the validation loop can deadlock.  Hence how to calculate the final and the intermediate metrics properly?

I don't think we should created uneven length datasets (with DistributedSampler subclass). Instead we could generate the same length datasets (same as PyTorch distributed sampler), and skip adding metrics from repeated samples (during the last epoch). This won't deadlock. Perhaps there is a better way?

**Comment by Nic-Ma** [comment]
Hi @myron ,

I think many original test datasets are uneven length, we should keep the same to get accurate metrics.
Anyway, maybe this is a good research topic to discuss, involve @wyli @ericspod for more discussion.

Thanks.

**Comment by myron** [comment]
yes, that's what I mean too. many original test datasets are  uneven length,  and we need to calculate the correct accuracy on multi-gpu.  it seems at the moment we are able to do it , but we can't get intermediate results to print, and the validation loop gives no metric feedback to the user...  so perhaps we can come up with a better way.  I'm not sure what this way is, but it should be a common problem in different domains (not only in medical)

**Comment by Nic-Ma** [comment]
Add @ericspod @rijobro python experts for more discussion.

Thanks in advance.

**Comment by myron** [comment]
> I think for this specific application we would need a utility for storing small values locally, perhaps using one of the key-value classes in `torch.distributed` where we would compute the metric locally and each process would upload the result keyed to its rank. Rank 0 would then sum/average whatever it finds regardless of count, so processes without anything to do can silently do nothing safely. This would need to be synchronised somehow without `Barrier` which would block on those idle processes.

@ericspod  Thank you for the reply. Yes, I think this would be an option. If we merge https://github.com/Project-MONAI/MONAI/pull/5383 
then we can use it to append data , but don't compute average on the last epoch (when not all processes participate). 

Alternative we could use evenly-divisible dataset (with repetition), but than skip results from repeated images to Metric average caluclation. This would require to track (for each image) if it's repeated or not inside of DistributedDataSampler

**Comment by JHancox** [comment]
@ericspod , @myron  - a bit late to the party, but facing a similar problem. In previous situations I have encountered, the last iteration batch size per epoch would either be rounded off to an n_gpu divisible number and the remaining (<n_gpu ) training examples would be skipped or, as @myron suggests, some inputs are, effectively, brought forward from the next epoch. Given that the training examples are shuffled between epochs, this should have no effect on the eventual convergence/accuracy. 
The alternatives of dealing with a sub-minibatch, such as padding and then masking out the empty results, all add a lot more code complexity for little, if any, practical gain. Interested to know your thoughts.

**Comment by ericspod** [comment]
@JHancox The solution during training depends on how you're sampling from the dataset of course, if you're using a sampler that pulls a certain number of samples (which is independent of the dataset size) then you can always get mini-batches of the right size. The issue with validation is that you want to validate each input exactly once unless you're doing some sort of test-time augmentation. We want to avoid unnecessary work as well but duplicating inputs on an undersized batch would work, provided you mask out the extra results from your final calculation. My alternative was to use some other mechanism to broadcast results such that idle GPUs can correctly do nothing but that was just for validation in my mind.

**Comment by JHancox** [comment]
Agreed @ericspod - validation does need to go that extra distance to ensure precision (unless we mandate all test sets to be of size [5040](https://en.wikipedia.org/wiki/5040_(number)) :) )
